### ISSUE TITLE
Simple fetch takes over 2 minutes (Android)

### ISSUE THREAD

**Issue by jackswiggett** [expo/expo#2756]
Making a simple GET request to an established website (e.g. Google or Wikipedia) is taking over 2 minutes on Android. I originally experienced this issue accessing my own API server in an ejected, pure React Native app (and it occurred in a production build of a beta app distributed via the Play Store, on multiple people's devices). But after a long time debugging I realized I could reproduce it in an extremely simple Expo app. In general it seems that the first request to a given website subdomain (e.g. `en.wikipedia.org`) takes an extremely long time, and subsequent requests to that subdomain execute very quickly.

This seems like such a basic use case that I can't believe it would be an issue with RN or Expo. But I'm at a loss for how to resolve it so here I am. Any ideas for how to debug this or information about whether you experience it would be much appreciated.

### Steps to reproduce

1. Go to https://snack.expo.io/ryZi8lGRm.
2. Click 'Run' and scan the QR code using the Expo app on an Android device.
3. Open the Logs panel in the snack editor.

I am testing using WiFi. My Pixel does not have a SIM card so I can't test over a cellular network. Accessing the same URLs in my phone's browser works as expected.

#### Device: Pixel, Android 9
The logs shown are from snack.expo.io. The first request takes over 2 minutes to complete. The last three lines of logs here (at `13:12:33`) are also unexpected; the app should only fetch each page once.
```
Pixel: 13:10:25 fetching https://en.wikipedia.org/wiki/Main_Page
Pixel: 13:12:31 fetched https://en.wikipedia.org/wiki/Main_Page
Pixel: 13:12:31 fetched https://en.wikipedia.org/wiki/React_(JavaScript_library)
Pixel: 13:12:31 fetching https://en.wikipedia.org/wiki/React_(JavaScript_library)
Pixel: 13:12:33 fetching https://en.wikipedia.org/wiki/React_(JavaScript_library)
Pixel: 13:12:33 fetched https://en.wikipedia.org/wiki/Main_Page
Pixel: 13:12:33 fetched https://en.wikipedia.org/wiki/React_(JavaScript_library)
```

#### Device: iPhone SE, iOS 12.1
The logs shown are from snack.expo.io. The requests both execute very quickly (although the first line of the logs here is unexpected).
```
Jack's iPhone: 1:40:22 PM fetched https://en.wikipedia.org/wiki/React_(JavaScript_library)
Jack's iPhone: 1:40:23 PM fetching https://en.wikipedia.org/wiki/Main_Page
Jack's iPhone: 1:40:23 PM fetched https://en.wikipedia.org/wiki/Main_Page
Jack's iPhone: 1:40:23 PM fetching https://en.wikipedia.org/wiki/React_(JavaScript_library)
Jack's iPhone: 1:40:23 PM fetched https://en.wikipedia.org/wiki/React_(JavaScript_library)
```

#### Device: Pixel API 23 emulator (Android Studio)
In this case, the logs didn't show up in snack.expo.io, but they showed up in Logcat in Android Studio. The first request took 16 seconds -- faster than 2 minutes, but still much slower than I'd expect.
```
11-20 14:05:30.528 29726-29856/host.exp.exponent I/ReactNativeJS: Running application "main" with appParams: {"initialProps":{"exp":{"initialUri":"exp://expo.io/@snack/ryZi8lGRm+S18T7lz07","manifest":{"iconUrl":"https://s3.amazonaws.com/exp-brand-assets/ExpoIcon_200.png","splash":{"image":"./splash.png","imageUrl":"https://d1wp6m56sqw74a.cloudfront.net/~assets/d76c685fe59a4e8a1e25a915e34f3222"},"releaseChannel":"default","primaryColor":"#1b73b4","orientation":"default","description":"No description","loadedFromCache":false,"revisionId":"1.0.0-r.AoKaPVJY9M","name":"slow network requests","commitTime":"2018-11-16T19:14:06.165Z","bundleUrl":"https://d1wp6m56sqw74a.cloudfront.net/%40exponent%2Fsnack%2F1.0.0%2F447447cdd7ef741b1c1b917698e8b7bb-31.0.0-android.js","notification":{"iconUrl":"https://s3.amazonaws.com/exp-us-standard/placeholder-push-icon-blue-circle.png","color":"#000000","icon":"https://s3.amazonaws.com/exp-us-standard/placeholder-push-icon-blue-circle.png"},"publishedTime":"2018-11-20T21:03:21.335Z","packagerOpts":{"assetExts":["ttf","otf","mp4"]},"sdkVersion":"31.0.0","id":"@snack/ryZi8lGRm","icon":"https://s3.amazonaws.com/exp-brand-assets/ExpoIcon_200.png","dependencies":{"react-native-paper":"2.1.3"},"slug":"snack","locales":{},"extra":{"dependencies":{"react-native-paper":{"version":"2.1.3","isUserSpecified":true}},"code":{"App.js":{"type":"CODE","contents":"import * as React from 'react';\nimport { Text, View } from 'react-native';\n\nconst getTimestamp = () => new Date().toLocaleTimeString();\n\nconst testFetch = async () => {\n  console.log(getTimestamp(), 'fetching https://en.wikipedia.org/wiki/Main_Page');\n  await fetch('https://en.wikipedia.org/wiki/Main_Page');\n  console.log(getTimestamp(), 'fetched https://en.wikipedia.org/wiki/Main_Page');\n\n  console.log(getTimestamp(), 'fetching https://en.wikipedia.org/wiki/React_(JavaScript_library)');\n  await fetch('https://en.wikipedia.org/wiki/React_(JavaScript_library)');\n  console.log(getTimestamp(), 'fetched https://en.wikipedia.org/wiki/React_(JavaScript_library)');\n};\n\nexport default class App extends React.Component {\n  componentDidMount() {\n    testFetch();\n  }\n\n  render() {\n    return (\n      <View><Text>{'Hello world'}</Text></View>\n    );\n  }\n}\n"},"README.md":{"type":"CODE","contents":"# Sample Snack app\n\nWelcome to Expo!\n\nOpen the `App.js` file to start writing some code. You can preview the changes directly on your phone or tablet by clicking the **Run** button or use the simulator by clicking **Tap to Play**. When you're done, click **Save** and share the link!\n\nWhen you're ready to see everything that Expo provides (or if you want to use your own editor) you can **Export** your project and use it with [expo-cli](https://docs.expo.io/versions/latest/introduction/installation.html).\n\nProjects created in Snack are publicly available, so you can easily share the link to this project via link, or embed it on a web page with the **Embed** button.\n\nIf you're having problems, you can tweet to us [@expo](https://twitter.com/expo) or ask in our [forums](https://forums.expo.io).\n\nSnack is Open Source. You can find the code on the [GitHub repo](https://github.com/expo/snack-web).\n"}}},"version":"1.0.0","hostUri":"expo.io/@snack/ryZi8lGRm","isVerified":true},"shell":false}},"rootTag":1}. __DEV__ === false, development-level warning are OFF, performance optimizations are ON
11-20 14:05:30.593 29726-29860/host.exp.exponent D/i: WARNING: getPackageName called on ScopedContext
11-20 14:05:30.660 29726-29860/host.exp.exponent D/i: WARNING: getPackageName called on ScopedContext
11-20 14:05:30.670 29726-29860/host.exp.exponent D/i: WARNING: getPackageName called on ScopedContext
11-20 14:05:30.672 29726-29860/host.exp.exponent D/i: WARNING: getPackageName called on ScopedContext
11-20 14:05:31.198 29726-29856/host.exp.exponent I/ReactNativeJS: '14:05:31', 'fetching https://en.wikipedia.org/wiki/Main_Page'
11-20 14:05:31.414 29726-29726/host.exp.exponent D/SoLoader: init exiting
11-20 14:05:47.513 29726-29856/host.exp.exponent I/ReactNativeJS: '14:05:47', 'fetched https://en.wikipedia.org/wiki/Main_Page'
    '14:05:47', 'fetching https://en.wikipedia.org/wiki/React_(JavaScript_library)'
11-20 14:05:47.606 29726-29856/host.exp.exponent I/ReactNativeJS: '14:05:47', 'fetched https://en.wikipedia.org/wiki/React_(JavaScript_library)'
11-20 14:05:54.533 1982-20315/com.google.android.gms.persistent E/WakeLock: release without a matched acquire!
11-20 14:05:54.558 2426-29877/com.google.android.gms I/EventLogChimeraService: Aggregate from 1542749753865 (log), 1542749753865 (data)
11-20 14:05:54.667 1616-1979/system_process I/AccountManagerService: getTypesVisibleToCaller: isPermitted? true
11-20 14:05:59.164 1235-1292/? D/hwcomposer: hw_composer sent 3155 syncs in 60s
```

### Environment

Not sure this is relevant, since I'm using snack.expo.io. But incidentally, when I ran `expo diagnostics` I got an error. Here's the output, with `EXPO_DEBUG=true`:

```
[13:53:12] Generating diagnostics report...
[13:53:12] You can join our slack here: https://slack.expo.io/.

  Expo CLI 2.4.0 environment info:
    System:
      OS: macOS High Sierra 10.13.6
      Shell: 3.2.57 - /bin/bash
    Binaries:
      Node: 8.9.1 - ~/.nvm/versions/node/v8.9.1/bin/node
      npm: 6.4.1 - ~/.nvm/versions/node/v8.9.1/bin/npm
      Watchman: 4.9.0 - /usr/local/bin/watchman
    IDEs:
      Android Studio: 3.1 AI-173.4697961
      Xcode: 10.1/10B61 - /usr/bin/xcodebuild
    npmPackages:
      expo: ^31.0.2 => 31.0.2 
      react: 16.5.0 => 16.5.0 
      react-native: https://github.com/expo/react-native/archive/sdk-31.0.0.tar.gz => 0.57.1 
    npmGlobalPackages:
      expo-cli: 2.4.0

  Diagnostics report:
|[13:53:17] Request failed with status code 500
[13:53:17] Error: Request failed with status code 500
    at createError (/Users/jackswiggett/.nvm/versions/node/v8.9.1/lib/node_modules/expo-cli/node_modules/axios/lib/core/createError.js:16:15)
    at settle (/Users/jackswiggett/.nvm/versions/node/v8.9.1/lib/node_modules/expo-cli/node_modules/axios/lib/core/settle.js:17:12)
    at IncomingMessage.handleStreamEnd (/Users/jackswiggett/.nvm/versions/node/v8.9.1/lib/node_modules/expo-cli/node_modules/axios/lib/adapters/http.js:236:11)
    at emitNone (events.js:111:20)
    at IncomingMessage.emit (events.js:208:7)
    at endReadableNT (_stream_readable.js:1056:12)
    at _combinedTickCallback (internal/process/next_tick.js:138:11)
    at process._tickCallback (internal/process/next_tick.js:180:9)
```

**Comment by jackswiggett** [comment]
I just tried this using my friend's Moto G 3 (scanning a QR code from snack.expo.io), and the same issue occurred:

```
MotoG3: 11:37:59 fetching https://en.wikipedia.org/wiki/Main_Page
MotoG3: 11:40:07 fetched https://en.wikipedia.org/wiki/Main_Page
MotoG3: 11:40:07 fetching https://en.wikipedia.org/wiki/React_(JavaScript_library)
MotoG3: 11:40:08 fetched https://en.wikipedia.org/wiki/React_(JavaScript_library)
```

Then I tried it using the Android device preview on snack.expo.io, and it worked as expected:

```
Android SDK built for x86: 15:19:35 fetching https://en.wikipedia.org/wiki/Main_Page
Android SDK built for x86: 15:19:35 fetched https://en.wikipedia.org/wiki/Main_Page
Android SDK built for x86: 15:19:35 fetching https://en.wikipedia.org/wiki/React_(JavaScript_library)
Android SDK built for x86: 15:19:35 fetched https://en.wikipedia.org/wiki/React_(JavaScript_library)
```

**Comment by jackswiggett** [comment]
Thanks for the followup @esamelson.

I originally tracked down this issue because I had a production app, distributed via Google Play, that was taking several minutes to load on some users' devices (including my own), and this was definitely happening on multiple different WiFi networks. I've seen the issue occur on my device on two different WiFi networks. That said, I just tried the following, and things appear to be working now:

1. Connected both my computer and Pixel to an iPhone personal hotspot. The issue went away.
2. Reconnected to WiFi. Then I was unable to even open the Expo app -- it froze on the white loading screen with the Expo logo.
3. Forgot the WiFi network on my Pixel, restarted the phone, restarted my router, reconnected to WiFi, then tried it again. The app worked without any problems.

I'm still concerned about this since it occurred on other networks/devices in the past, and since other apps on my Pixel haven't had any trouble connecting to WiFi. However, I understand that you may want to close this issue if you can't reproduce it.

One last request -- could you try running this app via Expo: https://expo.io/@kaimo/kaimo-app. Tap "log in" and enter `test@gmail.com` and `fakepassword`. This is a fake account but the app will need to make a request to determine that, so if you promptly see the "incorrect email/password" message, things are working as expected.

Thanks!

**Comment by esamelson** [comment]
Hi @jackswiggett - I tried this a couple of times, each time with a fresh install of the Expo client, and got the "incorrect email/password" message right away each time.

From what you wrote, it sounds like a problem between your device and the WiFi router. I'm going to go ahead and close this for now, but please feel free to ping me if you find out more information or a way to repro this more consistently and I'll reopen.

**Comment by shooka** [comment]
@jackswiggett We had a similar problem with the Android version of our Expo app. It turned out to be an IPv6 DNS record that somehow failed, and not related to Expo at all.

When we booted our app on Android using 4G, the first request to our API was extremely slow. When booting the Android app using 4G + VPN the request was suddenly fast, which led me to the solution. Apparently the default protocol is IPv6 on later versions of Android when connected to 4G, so I expected the VPN provider to not support IPv6. When dropping the IPv6 record everything ended up being as fast as the iOS counterpart within minutes.

At first, I expected it to be the first request, our app made, but we actually make a call to AWS to check if there are new versions of the app out, and if the user should be forced to upgrade. This request happened instantly which led me to believe that our API was the problem, but since the app hits different endpoints based on the user being logged in already vs. having to log in, this couldn't be it, and also once the first request was made, everything was smooth.

Anyway. We solved it. Hope this might be your problem too. It can happen on wifi or cellular based on if the entire path allows IPv6.
